src/core/geometry/geom3/retessellate.js

```diff
@@ -10,45 +9,43 @@ const fromPolygons = require('./fromPolygons')
   Polygons are split at each sweep line, and the fragments are joined horizontally and vertically into larger polygons
   (making sure that we will end up with convex polygons).
 */
-const retessellate = geometry => {
+const retessellate = (geometry) => {
+  if (!geometry.isCanonicalized) {
+    throw new Error('geometry must be canonical, call canonicalize first')
+  }
+
   if (geometry.isRetesselated) {
     return geometry
-  } else {
-    let polygonsPerPlane = {}
-    let isCanonicalized = geometry.isCanonicalized
-    let fuzzyfactory = new FuzzyFactory3d()
-    geometry.polygons.map( polygon => {
-      let plane = polygon.plane
-      let shared = polygon.shared
-      if (!isCanonicalized) {
-        // in order to identify polygons having the same plane, we need to canonicalize the planes
-        // We don't have to do a full canonizalization (including vertices), to save time only do the planes and the shared data:
-        plane = fuzzyfactory.getPlane(plane)
-        shared = fuzzyfactory.getPolygonShared(shared)
-      }
-      let tag = plane.getTag() + '/' + shared.getTag()
-      if (!(tag in polygonsPerPlane)) {
-        polygonsPerPlane[tag] = [polygon]
-      } else {
-        polygonsPerPlane[tag].push(polygon)
-      }
-    })
-    let destpolygons = []
-    for (let planetag in polygonsPerPlane) {
-      let sourcepolygons = polygonsPerPlane[planetag]
-      if (sourcepolygons.length < 2) {
-        destpolygons = destpolygons.concat(sourcepolygons)
-      } else {
-        const retesselayedpolygons = reTesselateCoplanarPolygons(sourcepolygons)
-        destpolygons = destpolygons.concat(retesselayedpolygons)
-      }
-    }
-    let result = fromPolygons(destpolygons)
-    result.isRetesselated = true
-    // result = result.canonicalized();
-    result.properties = geometry.properties // keep original properties
-    return result
   }
+
+  const polygonsPerPlane = new Map()
+  geometry.polygons.forEach((polygon) => {
+    let values = polygonsPerPlane.get(polygon.plane)
+    if (values === undefined) {
+      values = [polygon]
+      polygonsPerPlane.set(polygon.plane, values)
+    } else {
+      values.push(polygon)
+    }
+  })
+
+  let destpolygons = []
+  polygonsPerPlane.forEach((sourcepolygons) => {
+    if (sourcepolygons.length < 2) {
+      destpolygons = destpolygons.concat(sourcepolygons)
+    } else {
+      const retesselayedpolygons = reTesselateCoplanarPolygons(sourcepolygons)
+      destpolygons = destpolygons.concat(retesselayedpolygons)
+    }
+  })
+
+  const result = fromPolygons(destpolygons)
+  result.isCanonicalized = geometry.isCanonicalized
+  result.isRetesselated = true
+  // TODO result.properties = geometry.properties // keep original properties
```

Comment:
properties will be moved to the upper level (ie they are based on the transforms, not geometry, so geometries need not care :)